examples/step-40/step-40.cc

```diff
@@ -572,40 +572,11 @@ namespace Step40
 
     data_out.build_patches();
 
-    // The next step is to write this data to disk. We choose file names of
-    // the form <code>solution-XX.PPPP.vtu</code> where <code>XX</code>
-    // indicates the refinement cycle, <code>PPPP</code> refers to the
-    // processor number (enough for up to 10,000 processors, though we hope
-    // that nobody ever tries to generate this much data -- you would likely
-    // overflow all file system quotas), and <code>.vtu</code> indicates the
-    // XML-based Visualization Toolkit for Unstructured grids (VTU) file
-    // format.
-    const std::string filename =
-      ("solution-" + Utilities::int_to_string(cycle, 2) + "." +
-       Utilities::int_to_string(triangulation.locally_owned_subdomain(), 4));
-    std::ofstream output(filename + ".vtu");
-    data_out.write_vtu(output);
-
-    // The last step is to write a "master record" that lists for the
-    // visualization program the names of the various files that combined
-    // represents the graphical data for the entire domain. The
-    // DataOutBase::write_pvtu_record does this, and it needs a list of
-    // filenames that we create first. Note that only one processor needs to
-    // generate this file; we arbitrarily choose processor zero to take over
-    // this job.
-    if (Utilities::MPI::this_mpi_process(mpi_communicator) == 0)
-      {
-        std::vector<std::string> filenames;
-        for (unsigned int i = 0;
-             i < Utilities::MPI::n_mpi_processes(mpi_communicator);
-             ++i)
-          filenames.push_back("solution-" + Utilities::int_to_string(cycle, 2) +
-                              "." + Utilities::int_to_string(i, 4) + ".vtu");
-
-        std::ofstream master_output(
-          "solution-" + Utilities::int_to_string(cycle, 2) + ".pvtu");
-        data_out.write_pvtu_record(master_output, filenames);
-      }
+    // The next step is to write this data to disk. We write up to 8 VTU files
+    // in parallel with the help of MPI-IO. Additionally a PVTU record is
+    // generated, which groups the written VTU files.
+    data_out.write_vtu_with_pvtu_record(
+      "./", "solution", cycle, 2, mpi_communicator, 8);
```

Comment:
Is 8 a good number? We ran this program with up to 16k processors. I'd expect that with just 8 blocks we'd get into a rather bad bottleneck here!

Reply:
> just 8 blocks we'd get into a rather bad bottleneck here

It should not be a bottleneck. All 16k processes would write in parallel in 8 files via MPI-IO.

Reply:
There is several considerations:
1. Does writing the output scale? When I wrote the routines (a few years ago), I tested writing into a single output file from 1000+ cores and got excellent performance (10s of GB/s). I don't see a problem with this.
2. Is this sensible for visualization? If you do visualization in serial, it doesn't matter. If you visualize in parallel, it might be helpful to have more than one file. With 8, you can at least run on 8 nodes. Maybe it makes sense to increase this number slightly? Not that it matters that much...

Reply:
What if we used `max(8,ceil(n_procs/128))`? I mean, right now we write one file per processor; maybe we can find a compromise that at least makes sure we don't make this the one function that blocks everything if someone tried to run the program on 10k processors...